pkg/ibgu/ibgu.go

```diff
@@ -0,0 +1,331 @@
+)
+
+// IbguBuilder provides struct for the ibgu object containing connection to
+// the cluster and the ibgu definitions.
+type IbguBuilder struct {
+	// ibgu Definition, used to create the ibgu object.
+	Definition *v1alpha1.ImageBasedGroupUpgrade
+	// created ibgu object.
+	Object *v1alpha1.ImageBasedGroupUpgrade
+	// api client to interact with the cluster.
+	apiClient goclient.Client
+	// used to store latest error message upon defining or mutating application definition.
+	errorMsg string
+}
+
+// NewIbguBuilder creates a new instance of IbguBuilder.
+func NewIbguBuilder(
+	apiClient *clients.Settings,
+	name string,
+	nsname string) *IbguBuilder {
+	glog.V(100).Infof(
+		"Initializing new ibgu structure with the following params: name: %s, nsname: %s", name, nsname)
+
+	if apiClient == nil {
+		glog.V(100).Info("The apiClient for the ibgu is nil")
+
+		return nil
+	}
+
+	err := apiClient.AttachScheme(v1alpha1.AddToScheme)
+	if err != nil {
+		glog.V(100).Infof("Failed to add ibgu v1alpha1 scheme to client schemes")
+
+		return nil
+	}
+
+	builder := &IbguBuilder{
+		apiClient: apiClient.Client,
+		Definition: &v1alpha1.ImageBasedGroupUpgrade{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: nsname,
+			},
+			Spec: v1alpha1.ImageBasedGroupUpgradeSpec{
+				IBUSpec: lcav1.ImageBasedUpgradeSpec{},
+				Plan:    []v1alpha1.PlanItem{},
+			},
+		},
+	}
+
+	if name == "" {
+		glog.V(100).Infof("The name of the ibgu is empty")
+
+		builder.errorMsg = "ibgu 'name' cannot be empty"
+
+		return builder
+	}
+
+	if nsname == "" {
+		glog.V(100).Infof("The namespace of the ibgu is empty")
+
+		builder.errorMsg = "ibgu 'nsname' cannot be empty"
+
+		return builder
+	}
+
+	return builder
+}
+
+// WithClusterLabelSelectors appends labels to the ibgu clusterLabelSelectors.
+func (builder *IbguBuilder) WithClusterLabelSelectors(labels map[string]string) *IbguBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	if len(labels) == 0 {
+		glog.V(100).Infof("The 'labels' of the IBGU is empty")
+
+		builder.errorMsg = "can not apply empty cluster label selectors to the IBGU"
+
+		return builder
+	}
+
+	labelSelectors := []metav1.LabelSelector{
+		{
+			MatchLabels: labels,
+		},
+	}
+
+	builder.Definition.Spec.ClusterLabelSelectors = labelSelectors
+
+	return builder
+}
+
+// WithSeedImageRef appends the SeedImageRef to the ibuSpec.
+func (builder *IbguBuilder) WithSeedImageRef(seedImage string, seedVersion string) *IbguBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	if seedImage == "" {
+		glog.V(100).Info("The 'seedImage' parameter is empty")
+
+		builder.errorMsg = "seedImage cannot be empty"
+
+		return builder
+	}
+
+	if seedVersion == "" {
+		glog.V(100).Info("The 'seedVersion' parameter is empty")
+
+		builder.errorMsg = "seedVersion cannot be empty"
+
+		return builder
+	}
+
+	ibuSpec := lcav1.ImageBasedUpgradeSpec{
```

Comment:
ibguSpec ?

Reply:
It's actually ibuSpec as the ImageBasedGroupUpgrade spec contains an ibuSpec field based on the ImageBasedUpgradeSpec struct.